File: gym/dp/coin_change.py
# O(amount ^ |coins|)
def rec_min_coins(amount, coins):
    if amount == 0:
        return 0

    res = 2 << 32
    for coin in coins:
        if amount >= coin:
            res = min(res, rec_min_coins(amount - coin, coins) + 1)
    return res

# O(N^2)
def min_coins(amount, coins):
    dp = [0 for _ in range(amount + 1)]
    
    dp[0] = 0
    for i in range(1, amount + 1):
        dp[i] = 2 << 32
        for coin in coins:
            if i >= coin:
                dp[i] = min(dp[i], 1 + dp[i - coin])
    return dp[amount]

def main():
    [amount, ncoins] = list(map(int, input().split()))
    coins = list(map(int, input().split()))

    valid = 0
    for coin in coins:
        if amount % coin == 0:
            valid += 1

    if valid == 0:
        print("No solution")
    else:
        print(f"{rec_min_coins(amount, coins)} {min_coins(amount, coins)}")

if __name__ == "__main__":
    main()

File: lc/coreprep/topics/dp/CoinChange.java
import java.util.Arrays;
import java.util.Scanner;

public class CoinChange {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int amount = in.nextInt();
      int n = in.nextInt();
      int[] coins = new int[n];

      for (int i = 0; i < n; i++) {
        coins[i] = in.nextInt();
      }

      // validation
      int valid = 0;
      for (int coin : coins) {
        if (amount % coin == 0) {
          valid++;
        }
      }

      if (valid == 0) {
        System.out.println("No solution");
        return;
      }

      Arrays.sort(coins);

      for (int i = 0, j = n - 1; i < j; i++, j--) {
        int c = coins[i];
        coins[i] = coins[j];
        coins[j] = c;
      }

      System.out.printf("Amount = %d, Greedy = %d, Recursive = %d, DP = %d\n",
                        amount, greedyChange(amount, coins),
                        recursiveChange(amount, coins),
                        dpChange(amount, coins));
    }
  }

  private static int greedyChange(int amount, int[] coins) {
    int minChange = 0;

    while (amount > 0) {
      for (int coin : coins) {
        if (amount >= coin) {
          minChange++;
          amount -= coin;
          break;
        }
      }
    }

    return minChange;
  }

  private static int recursiveChange(int amount, int[] coins) {
    if (amount == 0) {
      return 0;
    }

    int minChange = Integer.MAX_VALUE;
    for (int coin : coins) {
      if (amount >= coin) {
        minChange =
            Math.min(minChange, 1 + recursiveChange(amount - coin, coins));
      }
    }

    return minChange;
  }

  private static int dpChange(int amount, int[] coins) {
    int[] dp = new int[amount + 1];

    dp[0] = 0;
    for (int i = 1; i <= amount; i++) {
      dp[i] = Integer.MAX_VALUE;

      for (int coin : coins) {
        if (i >= coin) {
          dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
        }
      }
    }

    return dp[amount];
  }
}